File: lib/src/main/java/io/github/goto1134/structurizr/export/d2/model/D2Object.java
package io.github.goto1134.structurizr.export.d2.model;

import com.structurizr.export.IndentingWriter;

import java.util.*;

public class D2Object {

    public static class Builder {

        public static final int STROKE_DASHED = 5;
        public static final int STROKE_DOTTED = 2;

        private final String name;
        private final List<D2Property<D2Keyword, ?>> properties = new ArrayList<>();
        private final List<D2Property<D2StyleKeyword, ?>> style = new ArrayList<>();

        public Builder(String name) {
            this.name = name;
        }

        public Builder label(String label) {
            properties.add(new D2WrappedStringProperty<>(D2Keyword.LABEL, label));
            return this;
        }

        public Builder shape(D2Shape shape) {
            properties.add(new D2Property<>(D2Keyword.SHAPE, shape));
            return this;
        }

        public Builder icon(Optional<String> icon) {
            icon.filter(it -> !it.isEmpty())
                    .ifPresent(it -> properties.add(new D2WrappedStringProperty<>(D2Keyword.ICON, it)));
            return this;
        }

        public Builder link(Optional<String> link) {
            link.filter(it -> !it.isEmpty())
                    .ifPresent(it -> properties.add(new D2WrappedStringProperty<>(D2Keyword.LINK, it)));
            return this;
        }

        public Builder tooltip(String tooltip) {
            properties.add(new D2WrappedStringProperty<>(D2Keyword.TOOLTIP, tooltip));
            return this;
        }

        public Builder fill(String fill) {
            style.add(new D2WrappedStringProperty<>(D2StyleKeyword.FILL_COLOR, fill));
            return this;
        }

        public Builder stroke(String stroke) {
            style.add(new D2WrappedStringProperty<>(D2StyleKeyword.STROKE_COLOR, stroke));
            return this;
        }

        public Builder strokeWidth(Optional<Integer> widthOptional) {
            if (widthOptional.isPresent()) {
                return strokeWidth(widthOptional.get());
            }
            return this;
        }

        public Builder strokeWidth(Integer width) {
            style.add(new D2Property<>(D2StyleKeyword.STROKE_WIDTH, width));
            return this;
        }

        public Builder strokeDash(Integer dash) {
            style.add(new D2Property<>(D2StyleKeyword.STROKE_DASH, dash));
            return this;
        }

        public Builder dashed() {
            return strokeDash(STROKE_DASHED);
        }

        public Builder dotted() {
            return strokeDash(STROKE_DOTTED);
        }

        public Builder multiple(boolean value) {
            style.add(new D2Property<>(D2StyleKeyword.SHAPE_MULTIPLE, value));
            return this;
        }

        public Builder fontColor(String fontColor) {
            style.add(new D2WrappedStringProperty<>(D2StyleKeyword.TEXT_FONT_COLOR, fontColor));
            return this;
        }

        public Builder fontSize(Integer fontSize) {
            style.add(new D2Property<>(D2StyleKeyword.TEXT_FONT_SIZE, fontSize));
            return this;
        }

        public Builder opacity(Double opacity) {
            style.add(new D2Property<>(D2StyleKeyword.OPACITY, opacity));
            return this;
        }

        public Builder withGroupStyle() {
            return fill("white").stroke("black");
        }

        public D2Object build() {
            return new D2Object(name, properties, style);
        }
    }

    public static Builder builder(String name) {
        return new Builder(name);
    }

    protected final String name;
    protected final List<D2Property<D2Keyword, ?>> properties;
    protected final List<D2Property<D2StyleKeyword, ?>> style;

    public D2Object(String name, List<D2Property<D2Keyword, ?>> properties, List<D2Property<D2StyleKeyword, ?>> style) {
        this.name = name;
        this.properties = Collections.unmodifiableList(properties);
        this.style = Collections.unmodifiableList(style);
    }

    public void startObject(IndentingWriter writer) {
        writeHeader(writer);
        writeProperties(writer);
    }

    public void writeHeader(IndentingWriter writer) {
        writer.writeLine(String.format("%s: {", name));
        writer.indent();
    }

    private void writeProperties(IndentingWriter writer) {
        properties.stream().sorted(Comparator.comparing(it -> it.getKeyword().toString())).forEach(p -> p.write(writer));
        writer.writeLine(String.format("%s: {", D2Keyword.STYLE));
        writer.indent();
        style.stream().sorted(Comparator.comparing(it -> it.getKeyword().toString())).forEach(p -> p.write(writer));
        endObject(writer);
    }

    public void writeObject(IndentingWriter writer) {
        startObject(writer);
        endObject(writer);
    }

    public static void endObject(IndentingWriter writer) {
        writer.outdent();
        writer.writeLine("}");
    }
}

File: lib/src/main/kotlin/io/github/goto1134/structurizr/export/d2/model/D2Object.kt
package io.github.goto1134.structurizr.export.d2.model

import com.structurizr.export.IndentingWriter


fun IndentingWriter.indented(block: IndentingWriter.() -> Unit) {
    indent()
    block()
    outdent()
}

/**
 * @param K D2 object property name
 * @param V D2 object property value
 */
open class D2Property<K, V>(@JvmField val keyword: K, @JvmField protected val value: V) {
    open fun write(writer: IndentingWriter) = writer.writeLine("$keyword: $value")
}

class D2WrappedStringProperty<K>(keyword: K, value: String) : D2Property<K, String>(keyword, value) {
    override fun write(writer: IndentingWriter) {
        writer.writeLine("$keyword: \"$value\"")
    }
}

sealed class D2Object(
    protected val properties: List<D2Property<D2Keyword, *>>, protected val style: List<D2Property<D2StyleKeyword, *>>
) {
    open fun writeObject(writer: IndentingWriter, writeInside: (IndentingWriter) -> Unit = { }) {
        properties.write(writer)
        if (style.isNotEmpty()) {
            writer.writeLine("${D2Keyword.STYLE}: {")
            writer.indented {
                style.write(writer)
            }
            writer.writeLine("}")
        }
        writeInside(writer)
    }

    private fun List<D2Property<*, *>>.write(writer: IndentingWriter) {
        for (property in sortedBy { it.keyword.toString() }) {
            property.write(writer)
        }
    }
}

class GlobalObject private constructor(
    properties: List<D2Property<D2Keyword, *>>, style: List<D2Property<D2StyleKeyword, *>>
) : D2Object(properties, style) {
    companion object {
        fun build(block: PropertyBuilder.() -> Unit = {}) = with(PropertyBuilder().apply(block)) {
            GlobalObject(properties, style)
        }
    }
}

open class NamedObject protected constructor(
    protected val name: String, properties: List<D2Property<D2Keyword, *>>, style: List<D2Property<D2StyleKeyword, *>>
) : D2Object(properties, style) {
    companion object {
        fun build(name: String, block: PropertyBuilder.() -> Unit = {}) = with(PropertyBuilder().apply(block)) {
            NamedObject(name, properties, style)
        }
    }

    open fun openObject(writer: IndentingWriter) {
        writer.writeLine("$name: {")
    }

    override fun writeObject(writer: IndentingWriter, writeInside: (IndentingWriter) -> Unit) {
        openObject(writer)
        writer.indented {
            super.writeObject(writer, writeInside)
        }
        writer.writeLine("}")
    }
}

class TextObject private constructor(
    name: String,
    private val language: String,
    private val text: String,
    properties: List<D2Property<D2Keyword, *>>,
    style: List<D2Property<D2StyleKeyword, *>>
) : NamedObject(name, properties, style) {

    companion object {
        fun build(name: String, language: String, text: String, block: PropertyBuilder.() -> Unit = {}) =
            with(PropertyBuilder().apply(block)) {
                TextObject(name, language, text, properties, style)
            }

        val UNICODE_NEWLINE_PATTERN = Regex("\\R")
    }

    override fun openObject(writer: IndentingWriter) {
        writer.writeLine("$name: |`$language")
        writer.indented {
            text.splitToSequence(UNICODE_NEWLINE_PATTERN).forEach(::writeLine)
        }
        writer.writeLine("`| {")
    }
}

class PropertyBuilder {
    companion object {
        const val STROKE_DASHED = 5
        const val STROKE_DOTTED = 2
    }

    internal val properties: MutableList<D2Property<D2Keyword, *>> = ArrayList()
    internal val style: MutableList<D2Property<D2StyleKeyword, *>> = ArrayList()

    fun label(label: String) = apply {
        properties.add(D2WrappedStringProperty(D2Keyword.LABEL, label))
    }

    fun shape(shape: D2Shape?) = apply {
        if (shape != null) properties.add(D2Property(D2Keyword.SHAPE, shape))
    }

    fun icon(icon: String?) = apply {
        if (!icon.isNullOrBlank()) properties.add(D2WrappedStringProperty(D2Keyword.ICON, icon))
    }

    fun link(link: String?) = apply {
        if (!link.isNullOrBlank()) properties.add(D2WrappedStringProperty(D2Keyword.LINK, link))
    }

    fun tooltip(tooltip: String?) = apply {
        if (!tooltip.isNullOrBlank()) properties.add(D2WrappedStringProperty(D2Keyword.TOOLTIP, tooltip))
    }

    fun fill(fill: String) = apply {
        style.add(D2WrappedStringProperty(D2StyleKeyword.FILL_COLOR, fill))
    }

    fun fillPattern(fillPattern: D2FillPattern?) = apply {
        if (fillPattern != null) style.add(D2Property(D2StyleKeyword.FILL_PATTERN, fillPattern))
    }

    fun direction(direction: D2Direction?) = apply {
        if (direction != null) properties.add(D2Property(D2Keyword.DIRECTION, direction))
    }

    fun stroke(stroke: String) = apply {
        style.add(D2WrappedStringProperty(D2StyleKeyword.STROKE_COLOR, stroke))
    }

    fun strokeWidth(width: Int?) = apply {
        if (width != null) style.add(D2Property(D2StyleKeyword.STROKE_WIDTH, width))
    }

    fun dashed() = strokeDash(STROKE_DASHED)
    fun dotted() = strokeDash(STROKE_DOTTED)
    @Suppress("MemberVisibilityCanBePrivate")
    fun strokeDash(dash: Int) = apply {
        style.add(D2Property(D2StyleKeyword.STROKE_DASH, dash))
    }

    fun multiple(value: Boolean) = apply {
        style.add(D2Property(D2StyleKeyword.SHAPE_MULTIPLE, value))
    }

    fun fontColor(fontColor: String) = apply {
        style.add(D2WrappedStringProperty(D2StyleKeyword.TEXT_FONT_COLOR, fontColor))
    }

    fun fontSize(fontSize: Int) = apply {
        style.add(D2Property(D2StyleKeyword.TEXT_FONT_SIZE, fontSize))
    }

    fun opacity(opacity: Double) = apply {
        style.add(D2Property(D2StyleKeyword.OPACITY, opacity))
    }

    fun animated(animated: Boolean) = apply {
        if (animated) style.add(D2Property(D2StyleKeyword.CONNECTION_ANIMATED, true))
    }

    fun near(near: D2NearConstant) = apply {
        properties.add(D2Property(D2Keyword.NEAR, near.toString()))
    }
}
